src/audio/selector.c

```diff
@@ -0,0 +1,445 @@
+ */
+static void sel_set_channel_values(struct comp_data *cd, 
+						uint32_t in_channels, uint32_t out_channels, uint32_t ch_idx)
+{
+	if ((in_channels != SEL_IN_2CH) || (in_channels != SEL_IN_4CH))
+	{
+		trace_selector_error("sel_set_channel_values() error: in_channels = %u",
+		in_channels);
+		cd->in_channels_count = SEL_IN_2CH;
+	}
+	
+	if ((out_channels != SEL_OUT_2CH) || (in_channels != SEL_OUT_4CH) ||
+	 (in_channels != SEL_OUT_1CH))
+	{
+		trace_selector_error("sel_set_channel_values() error: out_channels = %u",
+		in_channels);
+		cd->out_channels_count = SEL_OUT_2CH;
+	}
+	if (ch_idx > (SEL_IN_4CH - 1)){
+		trace_selector_error("sel_set_channel_values() error: ch_idx = %u",
+		in_channels);
+		cd->sel_channel = 0;
+	}
+
+}
+
+/**
+ * \brief Creates selector component.
+ * \param[in,out] data Selector base component device.
+ * \param[in] channel mask 
+ * \return Pointer to selector base component device.
+ */
+static struct comp_dev *selector_new(struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_selector *sel;
+	struct sof_ipc_comp_selector *ipc_sel =
+		(struct sof_ipc_comp_selector *)comp;
+	struct comp_data *cd;
+
+
+	trace_selector("selector_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_sel->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_SELECTOR, ipc_sel->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		COMP_SIZE(struct sof_ipc_comp_selector));
+	if (dev == NULL)
+		return NULL;
+
+	sel = (struct sof_ipc_comp_selector *)&dev->comp;
+	memcpy(sel, ipc_sel, sizeof(struct sof_ipc_comp_selector));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+	if (cd == NULL) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, cd);
+
+	/* verification of initial parameters */
+	sel_set_channel_values(cd, ipc_sel->input_channels_count,
+										ipc_sel->output_channels_count, ipc_sel->selected_channel);
+
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+/**
+ * \brief Frees selector component.
+ * \param[in,out] dev Selector base component device.
+ */
+static void selector_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_selector("selector_free()");
+
+	rfree(cd);
+	rfree(dev);
+}
+
+/**
+ * \brief Sets selector component audio stream parameters.
+ * \param[in,out] dev Selector base component device.
+ * \return Error code.
+ *
+ * All done in prepare() since we need to know source and sink component params.
+ */
+static int selector_params(struct comp_dev *dev)
+{
+	trace_selector("selector_params()");
+
+	/* all configuration work is done in prepare */
+
+	return 0;
+}
+
+
+/**
+ * \brief Sets Selector control command.
+ * \param[in,out] dev Selector base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int selector_ctrl_set_cmd(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata)
+{
+	//TODO add implementation aligned with with SW assumptions
```

Comment:
Use C style comments /* TODO: xxx */. Anyhow, I would advise on removing empty function form the initial commit.

Reply:
TODO style comments fixed,